File: problem70/Problem70.java
import java.util.*;
/**
 * Problem 70
 */

public class Problem70 {
	public static void main(String[] str) {
		int upperBound = (int)Math.sqrt(10000000);
		boolean[] prime = IntegerArithmetic.sieveEratosthenes(upperBound);
		int nbPrimes = 0;
		for (int i = 0; i < prime.length; i++) {
			if (prime[i]) nbPrimes++;
		}
		int[] primes = new int[nbPrimes];
		int j = 0;
		for (int i = 0; i < prime.length; i++) {
			if (prime[i]) {
				primes[j] = i;
				j++;
			}
		}	
		int index = 87109;
		float minRatio = (float)index / (float)phi(index, primes);
		for (int n = 2; n < 10000000; n++) {
			int phiOfn = phi(n, primes);
			if (Combinatorics.isPerm(phiOfn, n)) {
				float ratio = (float)n / (float)phiOfn;
				if (ratio < minRatio) {
					minRatio = ratio;
					index = n;
				} 
			}
		}
		System.out.println("The solution is " + index);
	}
	
	/**
	* Returns phi n
	*/
	public static int phi(int n, int[] primes) {
		ArrayList<ArrayList<Integer>> primeDec = IntegerArithmetic.primeDecomp(n, primes);
		int result = 1;
		for (int i = 0; i < primeDec.size(); i++) {
			result *= (primeDec.get(i).get(0) - 1) * (int)Math.pow(primeDec.get(i).get(0), primeDec.get(i).get(1) - 1);
		}
		return result;
	}
}

File: problem70/problem70.js
/**
 * Problem 70
 */

var IntegerArithmetic = require('../libraries/IntegerArithmetic.js');
var Combinatorics = require('../libraries/Combinatorics.js');
var upperBound = Math.floor(Math.sqrt(10000000));
var prime = IntegerArithmetic.sieveEratosthenes(upperBound);
var nbPrimes = 0;
for (var i = 0; i < prime.length; i++) {
	if (prime[i]) nbPrimes++;
}
var primes = [];
var j = 0;
for (var i = 0; i < prime.length; i++) {
	if (prime[i]) primes.push(i);
}	
var index = 87109;
var minRatio = index / phi(index, primes);
for (var n = 2; n < 10000000; n++) {
	var phiOfn = phi(n, primes);
	if (Combinatorics.isPerm(phiOfn, n)) {
		var ratio = n / phiOfn;
		if (ratio < minRatio) {
			minRatio = ratio;
			index = n;
		} 
	}
}
console.log('The solution is ' + index);

/**
* Returns phi n
* @param integer n
* @param array of integers primes
* @return integer
*/
function phi(n, primes) {
	var primeDec = IntegerArithmetic.primeDecomp(n, primes);
	var result = 1;
	for (var i = 0; i < primeDec.length; i++) {
		result *= (primeDec[i][0] - 1) * Math.floor(Math.pow(primeDec[i][0], primeDec[i][1] - 1));
	}
	return result;
}